### PR TITLE
Fix timing of circuit state changes

### LINKED ISSUE #7: Invalid ordering of errors in Reset() method
### Issue description

```
var latestError = _errorDates
    .OrderBy(x => x.Key)
    .Take(1)
    .FirstOrDefault();
```

Returns the oldest error causing a loop of state change: Open -> HalfOpen -> Open -> HalfOpen -> ... without any delay.

Change of ordering should fix the issue:
```
var latestError = _errorDates
    .OrderByDescending(x => x.Key)
    .Take(1)
    .FirstOrDefault();
```

### PR DESCRIPTION
**Pull request by gian-hancock**
Timing of circuit state changes is intended to be based on the time of the most recent error, however it is actually using the oldest error time.

Currently we can get behaviour where an error causes the circuit to Open, it waits `halfOpenPeriodInSeconds`, then transitions to HalfOpen - as expected. However after this the circuit rapidly "flip-flops" between Open and HalfOpen, seemingly ignoring `halfOpenPeriodInSeconds`.

Fixes #7

---
Rebus is [MIT-licensed](https://opensource.org/licenses/MIT). The code submitted in this pull request needs to carry the MIT license too. By leaving this text in, __I hereby acknowledge that the code submitted in the pull request has the MIT license and can be merged with the Rebus codebase__.

### CODE REVIEW

**Comment by mookid8000** [comment]
Hi @gian-hancock , do the tests pass on your machine?

**Comment by gian-hancock** [comment]
> Hi @gian-hancock , do the tests pass on your machine?

Hi, they do, I'm not quite sure what's going on in the AppVeyor build.

It may be because the test fails (intentionally) on commit de22698 which introduces the test before the subsequent commit 7aa4352 fixes it.

**Comment by gian-hancock** [comment]
> > Hi @gian-hancock , do the tests pass on your machine?
> 
> Hi, they do, I'm not quite sure what's going on in the AppVeyor build.
> 
> It may be because the test fails (intentionally) on commit [de22698](https://github.com/rebus-org/Rebus.CircuitBreaker/commit/de226988cf433720381f2af96699b6cbd7e39d79) which introduces the test before the subsequent commit [7aa4352](https://github.com/rebus-org/Rebus.CircuitBreaker/commit/7aa435235f01f6d3bbd200be5bf5f2f8b04153fd) fixes it.

Update on this: They fail on my machine when I run via the CLI `dotnet test -c Release --no-restore`. However they succeed in Visual Studio's Test Explorer.

I'll investigate more when I get the chance.

**Comment by gian-hancock** [comment]
@mookid8000 Progress update:

**TL;DR:** I'll rewrite the failing test (aiming for tomorrow AEDT), but I think it's uncovered something else which may be worth discussing.

I believe that failing test - which fails consistently in CI and intermittently on my machine - is picking up on some behaviour unrelated to this PR.

Specifically, there seems to be a race condition where sometimes the message which triggered the circuit to open gets retried before the circuit is closed (i.e. before workers is set to 0).

I'm fairly confident this is the case because I can get the test to fail consistently on my machine with the following change: https://github.com/gian-hancock/Rebus.CircuitBreaker/commit/5ee0500756da7b31b2823b79c98a0c4e352a6f4d

Here's some logs demonstrating this from the test `WaitHalfOpenPeriodBeforeHalfOpening`, run after making the above change (with some lines omitted for brevity):
```text
### [22:13:12.262] DeliveryCount: 1
[INF] Rebus.CircuitBreaker.MainCircuitBreaker (Rebus 1 worker 1): Circuit breaker changed from Closed to Open
[WRN] Rebus.CircuitBreaker.MainCircuitBreaker (Rebus 1 worker 1): ### Currently executing a message handler, setting workers asynchronously. Time: "22:13:12.270"
### [22:13:12.278] DeliveryCount: 2
### [22:13:12.279] DeliveryCount: 3	
### [22:13:12.279] DeliveryCount: 4
### [22:13:12.279] DeliveryCount: 5	
[WRN] Rebus.CircuitBreaker.MainCircuitBreaker (.NET TP Worker): ### Finished setting workers asynchronously. Time: "22:13:12.785"; Count 0

  Failed WaitHalfOpenPeriodBeforeHalfOpening(False) [10 s]
  Error Message:
     ### [22:13:22.070] Expect message delivery count to be '1' after circuit has transitioned from closed -> open
  Expected: 1
  But was:  5
```

I'll rewrite the `WaitHalfOpenPeriodBeforeHalfOpening` test introduced in this PR to be based on the timing circuit breaker state changes. This is a more direct test of the issue I was trying to address anyway.

If you think the above findings are worth addressing I can make a separate issue for that too.

**Comment by gian-hancock** [comment]
Hi @mookid8000, I've fixed the failing test as described [above](https://github.com/rebus-org/Rebus.CircuitBreaker/pull/10#issuecomment-2478655815).

Curious to hear your thoughts on merging this.

**Comment by mookid8000** [comment]
Excellent - thanks for contributing! 👍 
It's out as Rebus.CircuitBreaker 0.0.5 on NuGet.org now 🙂

**Comment by gian-hancock** [comment]
Thanks for your work on Rebus and for merging so quickly! 🙂
